src/Agent.Plugins/PipelineCache/FingerprintCreator.cs

```diff
@@ -159,25 +192,72 @@ internal static void CheckKeySegment(string keySegment)
             string filePathRoot,
             IEnumerable<string> keySegments)
         {
+            // Quickly validate all segments
+            foreach (string keySegment in keySegments)
+            {
+                CheckKeySegment(keySegment);
+            }
+
             var sha256 = new SHA256Managed();
 
             string defaultWorkingDirectory = context.Variables.GetValueOrDefault(
                 "system.defaultworkingdirectory" // Constants.Variables.System.DefaultWorkingDirectory
                 )?.Value;
 
             var resolvedSegments = new List<string>();
+            var exceptions = new List<Exception>();
 
-            foreach (string keySegment in keySegments)
+            Action<string, KeySegmentType, Object> LogKeySegment = (segment, type, details) =>
```

Comment:
>Object [](start = 43, length = 6)

MatchedFile[]?

Reply:
I used `Object` because I wanted the ability to support different types of details for different segment types. But, yah, currently `details` is only ever going to be a `MatchedFile[]`, but theoretically we might come up with a new type of segment in the future that supports a different type of detail (Foo). That was the thinking anyway ...